manager/role_manager.go

```diff
@@ -21,32 +31,67 @@ type roleManager struct {
 	raft     *raft.Node
 	doneChan chan struct{}
 
-	// pending contains changed nodes that have not yet been reconciled in
+	// pendingReconciliation contains changed nodes that have not yet been reconciled in
 	// the raft member list.
-	pending map[string]*api.Node
+	pendingReconciliation map[string]*api.Node
+
+	// pendingRemoval contains raft members, indexed by node ID, whose nodes have since
+	// been deleted - these members need to be removed from raft
+	pendingRemoval map[string]*api.RaftMember
+
+	// leave this nil except for tests which need to inject a fake time source
+	clocksource clock.Clock
 }
 
 // newRoleManager creates a new roleManager.
-func newRoleManager(store *store.MemoryStore, raftNode *raft.Node) *roleManager {
+func newRoleManager(store *store.MemoryStore, raftNode *raft.Node, clocksource clock.Clock) *roleManager {
 	ctx, cancel := context.WithCancel(context.Background())
 	return &roleManager{
-		ctx:      ctx,
-		cancel:   cancel,
-		store:    store,
-		raft:     raftNode,
-		doneChan: make(chan struct{}),
-		pending:  make(map[string]*api.Node),
+		ctx:                   ctx,
+		cancel:                cancel,
+		store:                 store,
+		raft:                  raftNode,
+		doneChan:              make(chan struct{}),
+		pendingReconciliation: make(map[string]*api.Node),
+		pendingRemoval:        make(map[string]*api.RaftMember),
+		clocksource:           clocksource,
 	}
 }
 
-// Run is roleManager's main loop.
-// ctx is only used for logging.
+// getTicker returns a ticker based on the configured clock source
+func (rm *roleManager) getTicker(interval time.Duration) clock.Ticker {
+	if rm.clocksource == nil {
+		return clock.NewClock().NewTicker(interval)
+	}
+	return rm.clocksource.NewTicker(interval)
+
+}
+
+// Run is roleManager's main loop.  On startup, it looks at every node object in the cluster and
+// attempts to reconcile the raft member list with all the nodes' desired roles.  If any nodes
+// need to be demoted or promoted, it will add them to a reconciliation queue, and if any raft
+// members' node have been deleted, it will add them to a removal queue.
+
+// These queues are processed immediately, and any nodes that failed to be processed are
+// processed again in the next reconciliation interval, so that nodes will hopefully eventually
+// be reconciled.  As node updates come in, any promotions or demotions are also added to the
+// reconciliation queue and reconciled.
+
+// The removal queue is never added to beyond the startup of the loop, because
```

Comment:
Thanks for adding this! 

>The removal queue is never added to beyond the startup of the loop, because before a node can be removed, it must be demoted, so the demotion should be sufficient to get it removed from the cluster membership.

Not sure I completely follow this part though :(

Reply:
What do you think about:

"Nothing will be added to the removal queue after the loop startup.  The removal queue consists of nodes that have been already removed from the cluster.  Before a manager (and hence raft) node can be removed from a cluster via the control API, it must first be demoted.  Because we watch for node updates, and we add demoted nodes to the reconciliation queue, the node's removal from the raft cluster should be handled upon reconciliation of that queue."

Reply:
Nm, as per https://github.com/docker/swarmkit/pull/2551#issuecomment-374801870, this comment won't apply any more anyway :|